feat: add Next.js 15+ support (continuation)

greg2012201 opened the pull request:
Hello! @andreizanik @joaopcm
I took over this PR #86  and successfully re-implemented smart imports to reduce dependency on `window`. 
We now check options to determine if the function is being called on the client or server. 
Additionally, we validate the presence of the `window` object to identify the render phase of the client component:
```ts
	
const getFn = (fnName: keyof typeof clientCookies, options?: OptionsType) => {
  if (isClientSide(options)) {
    if (getRenderPhase() === 'server') {
      // to prevent crash caused by missing window.document during server rendering phase
      return;
    }
    return clientCookies[fnName];
  }

  return serverCookies[fnName];
};

```
This step is necessary because, even if the function is designated for the client, Next.js attempts to use it during the server phase,
where objects like document are unavailable. This adjustment resolves environment-specific (server/client) errors.


I also encountered some issues with type compatibility. There was a type mismatch in NextRequest/NextResponse imported from `next/server`. 
These class declarations include an `[INTERNALS]` property of the `symbol` type, 
which was required by our library but missing in the Next.js application when exported from `next/server`.
Although I haven’t identified the root cause, defining custom interfaces has resolved the issue:

```ts 
interface NextCookiesRequest extends Request {
  get cookies(): RequestCookies;
}

interface NextCookiesResponse extends Response {
  get cookies(): ResponseCookies;
}

```


All tests have passed, and I have manually tested the functionality with Next.js v15.0.1. Everything appears to work correctly.

Please review the code and conduct your own tests to ensure readiness for release.

joaopcm reviewed (commented):
It is looking great to me!

I will stress test it to make sure it is properly handling all the cases as well! Great job!  🙌🏻

andreizanik commented on src/index.ts:
```diff
@@ -1,174 +1,41 @@
+
+const getFn = (fnName: keyof typeof clientCookies, options?: OptionsType) => {
+  if (isClientSide(options)) {
+    if (getRenderPhase() === 'server') {
```
```ts
import { getCookie } from 'cookies-next'

export const TestComponent = () => {
    const isActive = getCookie('active');
    
    return (
        <div>{isActive ? 'active': 'not active'}</div>
    );
}
```
it looks like we will have a difference between what the server and the client render. 

Because getCookie always returns undefined on the server phase.

andreizanik replied:
Oh, although the current version most likely has the same behavior

greg2012201 replied:
In my implementation, we simply return `undefined` when a function is designated for client-side use, and the client component calling it is being pre-rendered.

greg2012201 replied:
It's worth mentioning that `hasCookie` always returns a `boolean`. In a client component during the server phase, it returns `false` instead of `undefined.` In my opinion, having a consistent return type is important for `hasCookie`.

greg2012201 replied:
I see the issue now. One part of the problem is the mismatch between server and client. Additionally, if we use a function from the `cookies-next` import, we also encounter type-related issues:

Do you have any ideas on how to fix this? 

![image](https://github.com/user-attachments/assets/3417adcd-eafb-4a64-8a19-1a20e5e60a8d)

andreizanik commented on README.md:
````diff
@@ -3,393 +3,243 @@
+```
 
-## Usage
+For Next.js versions 12.2.0 to 13.x, use cookies-next version 4.3.0:
````
```suggestion
For Next.js versions 12.2.0 to 14.x, use cookies-next version 4.3.0:
```

andreizanik commented:
@greg2012201 Thanks!

> I will stress test it to make sure it is properly handling all the cases as well!

@joaopcm do you have any results?

greg2012201 commented:
> Thank you all for this it looks great :)
> 
> However, I ran into this error while using it in my project:
> 
> ```
> Error: Switched to client rendering because the server rendering errored:
> 
> document is not defined
> ```
> 
> This happens for me when using `getCookie` from `cookies-next/client` in a client component.

@ck-euan You're absolutely right; I missed that and forgot to add the render phase check in the client-side functions. I've pushed a fix for this.

Please test my changes again. I believe we're almost done.

greg2012201 commented:
After my changes, we still encounter issues. Please refer to: https://github.com/andreizanik/cookies-next/pull/87#discussion_r1824467776

I'm considering whether we should implement hooks for the client side. We could maintain the old functions in asynchronous versions for the server side, along with hooks for the client side, all exported from a single `index.ts` file.

What do you guys think?

andreizanik commented:
> After my changes, we still encounter issues. Please refer to: https://github.com/andreizanik/cookies-next/pull/87#discussion_r1824467776

I'm considering whether we should implement hooks for the client side. We could maintain the old functions in asynchronous versions for the server side, along with hooks for the client side, all exported from a single index.ts file.

What do you guys think? 

May be in next PR? 

This PR looks good to me

greg2012201 commented:
@andreizanik I've added the last two commits to this PR. I updated the export functions to narrow down return types and updated the documentation to explain how to use cookies-next in client components without causing mismatch errors.

Please check if everything works for you!

andreizanik commented:
Today I published a new version. 
Thanks @joaopcm @greg2012201 and everyone who was involved

bombillazo commented:
How can we make this work with Next.js Pages router as it was in v4?
